Add Tutorial for LOJ-1077 How Many Points? (en) (#454)
lightoj-dev/problem-tutorials: 1077/en.md

# LOJ 1077 - How Many Points?

## Solution

> The Number of lattice points lie on the segment **A(x1, y1) -> B(x2, y2)** is same for segment **A(0, 0) -> B(x2 - x1, y2 - y1)** (co-ordinate translation doesn't change the relative distance of lattice points from each other) and so for segment **A(0, 0) -> B(|x2 - x1|, |y2 - y1|)** (sign (+/-) only let's us know in which quadrant the segment will fall. Number of lattice points is length dependent not quadrant dependent).

Reviewer: ```suggestion
The Number of lattice points lie on the segment **A(x1, y1) -> B(x2, y2)** is same for segment **A(0, 0) -> B(x2 - x1, y2 - y1)** because co-ordinate translation doesn't change the relative distance of lattice points from each other.
The number of lattice points will also be equal to that of segment **A(0, 0) -> B(|x2 - x1|, |y2 - y1|)** since sign (+/-) only tells us in which quadrants the segment will fall.
Number of lattice points is length dependent not quadrant dependent.
```

So, how to calculate the number of lattice points on any segment **A(0, 0) -> B(x, y)** where **x, y >= 0**? The answer is: **gcd(x, y) + 1**. Why?

Suppose, **g = gcd(x, y)** then, all the lattice points are:

**(0 * x/g, 0 * y/g), (1 * x/g, 1 * y/g), . . . . . . . . . . . . . . . , ((g-1) * x/g, (g-1) * y/g), (g * x/g, g * y/g)** total of **(g + 1)** points with integer abscissas and ordinates.

But what's the proof they lie on the segment **AB** and there can't be any other lattice points? 

See: https://math.stackexchange.com/questions/628117/how-to-count-lattice-points-on-a-line

## Complexity

> - Time Complexity: O(T).

Reviewer: GCD has a logarithmic complexity, doesn't it?

Author: Yes, I skipped over that. Now fixed!

- Memory Complexity: O(1).

## Code

### C++

```cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long ll;
typedef pair <ll, ll> pii;

> 

Reviewer: Let's get rid of the definitions unless they are absolutely necessary.

## Solution

> The Number of lattice points lie on the segment **A(x1, y1) -> B(x2, y2)** is same for segment **A(0, 0) -> B(x2 - x1, y2 - y1)** because co-ordinate translation doesn't change the relative distance of lattice points from each other.

Reviewer: ~~The Number of lattice points lie~~

The number of lattice points lying on

Author: done!

Reviewer: > The number of lattice points **lying** on

You missed one.

Reviewer: We can merge right after this one. @Talha-Taki002

Author: I am so sorry for my overlooking mistake

Reviewer: Yeah no worries. It's not a biggie. I was just nitpicking there.

See: https://math.stackexchange.com/questions/628117/how-to-count-lattice-points-on-a-line

## Complexity

> - Time Complexity: O(T * lg(N)). Where N = **max(a, b)** of **gcd(a, b)**. [Check](https://www.quora.com/What-is-the-time-complexity-of-Euclids-GCD-algorithm/answer/M00nlight-2?ch=10&oid=16747157&share=a62b29f9&srid=xek7a&target_type=answer) for the time complexity of Eulclid's GCD algorithm.

Reviewer: Let's not point to Quora for references. They often lock answers. This might work as well: https://stackoverflow.com/questions/3980416/time-complexity-of-euclids-algorithm.

Additionally, typo on "Euclid".

Author: done!
